### ISSUE TITLE
Improve EDL

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/cloud#523]
https://github.com/PaddlePaddle/cloud/blob/develop/doc/design/scheduler.md

**Comment by gongweibao** [comment]
Maybe we don't need to make a big change to Kubernetes `default-scheduler`:
### About Priority and Preemption:
Kubernetes 1.9 support [Pod Priority and Preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/)
  - We can define [PrioriyClass](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass) ourself.
  - [PDB](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#poddisruptionbudget-is-supported-but-not-guaranteed) is supported.
  - High Priority Pod will be [placed ahead of](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order) other pending Pods on scheduling order.
  - Lowest Priority pod will be tried to [remove first](https://github.com/kubernetes/kubernetes/blob/release-1.9/plugin/pkg/scheduler/core/generic_scheduler.go#L643) when preemption happens.
 
### About `GPU required may have higher priority than CPU required pods`.
- We can use [Pod Affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) to try place pods with desired nodes but not guaranteed.
- Kubernetes support only high priority pods preempt lower priority pods and we don't always desire to set higher priority to `GPU required pods`.
    - We can schedule these pods first in a **time-window** when many pods pends in Kubernetes pods-scheduler-queue.It's not supported by `default-scheduler` now.

### About preemption of a pod.
In our V2 version EDL, `TrainingJob` resource has three type components `master` `pserver` `trainers` and we set `min-instance` and `max-instance` to `trainers`.
Since Kubernetes deploy pod one by one and it doesn't care about what type jobs it belongs now. When one pod is pending of `TrainingJob`:
- If it's `master` or `pserver`, the whole user job will be pending and the utilization of cluster will be down.
- If it's trainer, if trainer's number is smaller than `min-instance`, the pods' utilization will be down.

So we may need to support:
- In our controller scale down `TrainingJob` workers by custom metrics resource occupancy rate such as GPU not only by cluster pods occupancy rate. And then
- **Delete all pods of `TrainingJob`** rather than delete one pod of it.The `default-scheduler` doesn't support it now.


### Reference:
- [Multi-Scheduler in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/multiple-schedulers.md)
- [Pod Priority and Preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order)
- [Schedule a group of pods all at once](https://github.com/kubernetes/kubernetes/issues/16845)
- [The evolution of cluster scheduler architectures](http://www.firmament.io/blog/scheduler-architectures.html)

**Comment by typhoonzero** [comment]
Pod priority seems may not be our choices because:

1. In Kubernetes 1.9, it's still alpha.
1. It only considers static priority of `Pending` pods, but in [our design](https://github.com/PaddlePaddle/cloud/blob/develop/doc/design/scheduler.md) the priority is dynamic since it uses CFS.
1. Jobs require GPU (or other hardware accelerators) need to get even higher priority, e.g. if a CPU job has higher priority and GPU is still free, but the GPU job don't have enough CPU to run, we should let the GPU job run in advance.

I think Kubernetes Pod Priority is an excellent feature that let general applications to have their priority. PaddlePaddle's scheduler can run as a secondary scheduler (as you listed in the references) in the cluster to achieve our goal and don't affect the original one.

**Comment by gongweibao** [comment]
[How pod preemption happen between multi schedulers?](https://github.com/kubernetes/kubernetes/issues/57315#issuecomment-358566135?)

**Comment by resouer** [comment]
> PaddlePaddle's scheduler can run as a secondary scheduler (as you listed in the references) in the cluster to achieve our goal and don't affect the original one.

Is PaddlePaddle's scheduler have cached info of pod, node etc? Current multi-scheduler in Kubernetes will not handle cache inconsistency. You may want to use customized scheduler in that case.

**Comment by typhoonzero** [comment]
> Is PaddlePaddle's scheduler have cached info of pod, node etc?

No, not for now. Thanks for reminding, we'll take this into future designs.
